src/batSplay.ml

```diff
@@ -314,6 +314,70 @@ struct
     try find k m
     with Not_found -> def
 
+  (* option 1: simple implementation, dummy "find" invocation to perform rebalancing *)
+  let find_first_opt f (map : 'a t) =
+    let rec loop found tr =
+      match tr with
+      | Node(l, (k, v), r)
+        -> if f k
+           then loop (Some (k, v)) l
+           else loop found r
+      | Empty -> found in
+    let ret = loop None (sget map) in
+    (* dummy find to rebalance the tree *)
+    if Option.is_some ret
+    then ignore(find (fst (Option.get ret)) map);
+    ret
+
+  (* option 2: no unneccessary "find" invocation (saves several comparisons),
+     but implementation is rather complicated *)
```

Comment:
This implementation is too complex. I haven't looked at the details, but my intuition would be to write a `cfind_first` function.